### PR TITLE
RFC: High level Keras API for TF-Java

### PR DESCRIPTION
**Pull request by dhruvrajan**
Please see sigs/jvm/20191023-tensorflow-keras.md for a description of the proposed high-level Keras API. Would love to get feedback and comments; hopefully we can work towards an API that is convenient for users.

### CODE REVIEW

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+integrated into existing projects.
+
+### Overview
+
+Keras is built around a few core abstractions which comprise much of what is needed to build
+deep learning models. These include layers, models, optimizers, activations, losses, metrics,
+regularizers, and initializers.
+
+To make our code clearer and more manageable, we distinguish explicitly between two steps:
+Graph building (even in eager mode, graph "definition"), and Graph execution.
```
Comment (karllessard) [inline sigs/jvm/rfcs/20191023-tensorflow-keras.md:54]: Should we differentiate the different  graph execution environments? i.e. `Session` vs `EagerSession`? How the selected environment will impact the Keras API?

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    // within this method, call this.getDtype()
+    protected abstract void build(Ops tf, Shape inputShape)
+
+    // Call this method when using a created Layer object externally.
+    // Requires the user to provide a dtype, as some tensorflow 
+    // operations require the explicit specification of a dtype
+    // (e.g. tf.variable(Shape shape, Class<T> dtype) and tf.placeholder(Class<T> dtype))
+    public final void build(Ops tf, Shape inputShape, Class<T> dtype)
+    ```
+2. `Layer.call(Ops tf, Operand<T>... inputs)`
```
Comment (karllessard) [inline sigs/jvm/rfcs/20191023-tensorflow-keras.md:97]: I'm wondering if there is a gain to pass the `Ops` instance to every call of the `Layer` interface. Since `Layer` already seems to be a stateful API, can't we store the `Ops` that was use for creating it and reuse on subsequent call?

Also, maybe we could avoid to pass it at all by letting the Keras API use the default eager environment unless specified otherwise in some initializer, something like `Keras.initEnvironment(Ops tf)`? I'm not sure there is a gain for the user to keep direct access to the `Ops` instance.

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+on each other to build neural networks. Thus they must encode *differentiable* functions (i.e. built
+using TF Operations).
+
+Layers are defined by (1) a set of weights, and (2) a differentiable "transform" function which uses
+these weights to transform input tensor(s). For example, a `Dense` layer contains a kernel
+weight matrix *w_k* and a bias weight matrix *w_b*, and transforms an input *x* using the
+function *x* -> *f*(*w_k* x + *w_b*) for a user-specified activaition *f*.
+
+#### Defining Layers
+To define a new type of layer, users can override 3 methods from the base `Layer<T>` class.
```
Comment (karllessard) [inline sigs/jvm/rfcs/20191023-tensorflow-keras.md:69]: Just as a proposal here opened to discussion, instead of relying on inheritance for customizing layers, could we use composition and a factory? something like
```java
public interface LayerBackend<T> {

    void build(Ops tf, Shape inputShape);

    Operand<T> call(Ops tf, Operand<T>... operands);
}

final class DenseLayerBackend<T> implements LayerBackend<T> {
   ...
}

final class Layers {

    public static Layer<T> create(Shape shape, DataType<T> dtype, LayerBackend<T> backend) {
        return Layer.build(shape, dtype, backend);
    }

    public static Layer<T> createDense(Shape shape, DataType<T> dtype) {
        return layerOf(shape, dtype, new DenseLayerBackend<T>());
    }
}
```

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+Standard and user-defined layers might require arguments that are optional, and have defaults.
+To allow for this, the standard layer types each define `Options` and `Builder` classes which allow
+the specification of default and optional arguments. Since using an `Options` class for
+every layer can be cumbersome, static helper functions for creating standard layers
+are defined in the `Layers` class.
+
+Creating a `Dense` layer:
+```java
+// Using Layer Options Builder
+Layer<Float> dense = new Dense<>(128, Dense.Options.builder()
```
Comment (karllessard) [inline sigs/jvm/rfcs/20191023-tensorflow-keras.md:130]: How is the `Float` inferred here? and is `128` a shape? If so, I guess it should be `Shape.make(128)` unless you want to override the `Shape` class in TF Tools. And where is passed the `Ops` instance here, as we saw previously?

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+    .setActivation(Activations.relu)
+    .setKernelInitializer(Initializers.randomNormal)
+    .setBiasInitializer(Initializers.zeros)
+    .build())
+
+// Using static helper
+Layer<Float> dense = Layers.dense(128, Activations.relu, Initializers.randomNormal, Initializers.zeros)
```
Comment (karllessard) [inline sigs/jvm/rfcs/20191023-tensorflow-keras.md:137]: This is helpful for standard layers but it sounds like any user-defined layer (which should be pretty common I guess?) won't be able to reuse those factories and option builders. I'm wondering if we can improve this.

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+```
+
+
+## Detailed Design
+
+### Managing the Graph and Ops Accessors
+To use TensorFlow, we must have access to a `Graph` to build and run computations. In Python, this is constructed implicitly. In Java, the `Graph` and `Ops` objects must be created explicitly. Thus, we
```
Comment (karllessard) [inline sigs/jvm/rfcs/20191023-tensorflow-keras.md:144]: or `EagerSession`... or use the default, which is implicitly defined and used if you simply invoke `Ops.create()`

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+When using the Keras API, models are "compiled" and "fitted" within this block, and the `tf` object is passed through the internal method calls.
+
+### Keras Identifiers
+Standard Keras objects are very often referenced just by name, and don't all have
+to be created explicitly by the user.
+For example, in Python, the strings `"sgd"` or `"adam"` can be used to create different
+kinds of `Optimizer`s, `"relu"` and `"softmax"` stand for specific `Activations`, etc.
+
+In Java, the supported standard types are enumerated in various enum classes: `Activations`, 
+`Intializers`, `Optimizers`, etc., so these types become `Optimizers.sgd`, `Activations.relu`,
```
Comment (karllessard) [inline sigs/jvm/rfcs/20191023-tensorflow-keras.md:165]: Java standard is to have uppercase labels in enum, e.g. `Activations.RELU`, etc.

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+
+Each of these classes has a `select` method which convers the enum value to an object of the
+target type. For example, `Optimizers.select(Optimizers.sgd)` returns an `Optimizer`. However
+the end users shouldn't need to interact with this.
+
+### Keras Configuration
+Currently, the `~/.keras` directory is used for storing downloaded files, and user configurations 
+(for now, just datafiles for standard datasets). This works well when the Java keras client is used
+in isolation, but potentially could cause conflicts when a user is (likely) using the Python client
+for other projects.
```
Comment (karllessard) [inline sigs/jvm/rfcs/20191023-tensorflow-keras.md:177]: can we store potentially-conflicting file in a subfolder of `~/.keras` then?

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+        Arg1 arg1;
+        Arg2 arg2;
+        
+        // Create KerasObject from Options class
+        KerasObject kerasObject =
+            new KerasObject(KerasObject.Options.builder()           
+                .setArg1(arg1)
+                .setArg2(arg2)
+                .build());
+    }
```
Comment (karllessard) [inline sigs/jvm/rfcs/20191023-tensorflow-keras.md:206]: Not sure to understand this code... where is initialized `arg1` and `arg2`? and what the `kerasObject` is used for?

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+            Layers.flatten(),
+            Layers.dense(128, Activations.softmax, Initializers.randomNormal, Initializers.zeros),
+            Layers.dense(10, Activations.softmax, Initializers.randomNormal, Initializers.zeros)
+        );
+    }
+
+    public static Model<Float> train(Model<Float> model) throws Exception {
+        try (Graph graph = new Graph()) {
+            Ops tf = Ops.create(graph);
+            model.compile(tf, Optimizers.sgd, Losses.sparseCategoricalCrossentropy, Metrics.accuracy);
```
Comment (karllessard) [inline sigs/jvm/rfcs/20191023-tensorflow-keras.md:287]: by curiosity, what is the signature of `model.compile`?  Is there only one possible variant?

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+
+            Pair<GraphLoader<Float>, GraphLoader<Float>> loaders = MNIST.graphLoaders2D();
+            // GraphLoader objects contain AutoCloseable `Tensor` objects.
+            try (GraphLoader<Float> train = loaders.first(); GraphLoader<Float> test = loaders.second()) {
+                model.fit(tf, train, test, 10, 100);
```
Comment (karllessard) [inline sigs/jvm/rfcs/20191023-tensorflow-keras.md:292]: and what is the signature of `model.fit`?

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+
+    static {
+        model = Sequential.of(
+            Float.class, // dtype which will be passed through layer.build
+            Layers.input(28, 28),
+            Layers.flatten(),
+            Layers.dense(128, Activations.softmax, Initializers.randomNormal, Initializers.zeros),
+            Layers.dense(10, Activations.softmax, Initializers.randomNormal, Initializers.zeros)
+        );
+    }
```
Comment (karllessard) [inline sigs/jvm/rfcs/20191023-tensorflow-keras.md:282]: This looks very clean! It would be nice if we can help achieving a similar result when using a user-defined layers.

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+    classOf[JFloat],
+    input(28, 28),
+    flatten(),
+    dense(128, activation = relu, kernelInitializer = randomNormal, biasInitializer = zeros),
+    dense(10, activation = softmax, kernelInitializer = randomNormal, biasInitializer = zeros)
+  )
+
+  def train(model: Model[JFloat]): Model[JFloat] = {
+      Using.resource(new Graph()) { graph => {
+        implicit val tf: Ops = Ops.create(graph)
```
Comment (karllessard) [inline sigs/jvm/rfcs/20191023-tensorflow-keras.md:320]: I see, `tf` is implicit with Scala, but afaik it is not possible to achieve this in any other JVM languages, so we should probably optimize the API for the most common case.

**Review by karllessard — COMMENTED** [review]
Hey @dhruvrajan , very interesting RFC indeed, I've dropped a few comments/questions to start the discussions. There is a few typos that should be fixed as well but we'll keep that for the final version.

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md (hunk 2 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20191023-tensorflow-keras.md]: I have experimented with a few options for managing this. The available options seem to be

1. Leave the management of `ExecutionEnvironment` and `Ops` instances completely up to the user; on the API side, this means accepting `tf` or `ExecutionEnvironment` as a function argument anywhere it is needed.

Comment (dhruvrajan) [reply sigs/jvm/rfcs/20191023-tensorflow-keras.md:97]: Good points here, Karl, thanks! I actually experimented with a couple different approaches like you outlined.

 My feeling recently is that management of `ExecutionEnvironment` and `Ops` might actually be something to design / solve at the `TensorFlow` level, rather just at the `Keras` level, since this is a usage pattern every user of TF-Java will have to manage, even if not using a sub-framework like Keras.

The options available, as far as I can tell, are to

(1) Leave the management of these objects completely to the user. On the framework side, this means exposing `ExecutionEnvironment` and `Ops` instances as arguments to user-callable functions when needed.

This is the approach I took so far. I definitely agree with your point regarding `Layer`.  Passing in `Ops` only once to `Layer.build`, and storing a reference for re-use during `Layer.call` is a good idea, since you're never supposed to call `build` and `call` with different `ExecutionEnvironment` objects.

(2) Manage a global `ExecutionEnvironment` statically (maybe in the `TensorFlow` class). This way, even within internal framework methods, you can always get the current `ExecutionEnvironment` or `Ops` instance with a call like `TensorFlow.getCurrentEnvironment()` or `TensorFlow.ops()`. 

A strength, perhaps, of TF-Java right now is that you can work with multiple `ExecutionEnvironment` objects at the same time; this solution wouldn't really allow this, which is why I went with the first solution. admittedly, there's probably not a ton of use for this, but given that it exists, we can enable such use?

(3) Simulate the a Python session context manager.  A Pythonic way of handling this would be to create an internal "stack" of `ExecutionEnvironment` objects which are referenced by a context manager, so you can have something like:

```
with EagerSession() as env1:
   ... {TensorFlow code using env1}
   with EagerSession() as env2:
       ... {TensorFlow code using env2}
   ... {TensorFlow code using env1}
```

In Java, there are a couple ways I could see of emulating these semantics:

1. Using a try-with-resources and wrapper methods for generating `ExecutionEnvironment Objects, you could do something like the following:

```
class TensorFlow {
   private static Stack<ExecutionEnvironment> environments = new Stack();
   
   public static EagerSession getEagerSession() {
      environments.push(new EagerSession());
      return environments.top();
   }

   public static Graph getGraph() {
      environments.push(new Graph());
      return environments.top();
   }

   public static Ops ops() {
      return Ops.create(environments.top());
   }
}

// On EagerSession.close, call TensorFlow.environments.pop().close()
try (EagerSession eagerSession1 = TensorFlow.getEagerSession()) {
   try (EagerSession eagerSession2 = TensorFlow.getEagerSession()) {
      Ops tf = TensorFlow.ops(); // using eagerSession2
      ...
      
   } // Pop & Close eagerSession2
   
   Ops tf = TensorFlow.ops(); // using eagerSession1
   ...

} // Pop & Close eagerSession1
```

2. Or, we could use a wrapped function called something like `withEnvironment`, which takes an `ExecutionEnvironment`, and a `Lambda` accepting an `Ops` instance, which runs the lambda with this context; composing these functions could approximate the Python semantics
```
withEnvironment(new EagerSession(), tf -> {
   withEnvironment(new EagerSession(), tf -> { ... using inner env })
   ... using outer env
})
```

Should we add something like this to TensorFlow? Would love to know your thoughts.

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md (hunk 3 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20191023-tensorflow-keras.md:69]: I definitely like the idea of the `Layers` factory methods. I do see where you're coming from in terms of composition: subclassing layers is very much (likely) about that. (Please correct me if I'm missing something about your intent.)

But many Python layer implementations I've seen subclass mainly to share internal Layer state.

The "Model/Layer [Subclassing API](https://www.tensorflow.org/guide/keras/custom_layers_and_models)" is by now widely understood and used. Since that API is standard, and the subclassing semantics are simple, my inclination is to stick with it here. This makes it easier for users who port custom layer implementations from python, with deep class hierarchies. Thoughts?

File: sigs/jvm/rfcs/20191023-tensorflow-keras.md (hunk 9 above)
Comment (dhruvrajan) [reply sigs/jvm/rfcs/20191023-tensorflow-keras.md:206]: This is just an example structure for any Object we want to define that takes multiple optional/default parameters; hoping to standardize this structure across `Layer`, `Optimizer`, `Initializer`... arg1 and arg2 are null here until initialized via a builder call to `setArg1` or `setArg2`
